File: go/luhn/luhn.go
package luhn

import (
	"regexp"
	"unicode"
)

// Returns whether a luhn code is valid or not.
func Valid(id string) bool {
	if containsInvalidRune(id) {
		return false
	}

	digits := onlyDigits(id)

	if isInvalidDigits(digits) {
		return false
	}

	sum := luhnSum(digits)

	return isValidLuhnSum(sum)
}

// Returns whether or not a string contains any runes that cannot be
// present in a luhn code.
func containsInvalidRune(id string) bool {
	return invalidRunePattern.MatchString(id)
}

// Returns whether or not a list of digits has a chance of being a valid
// Luhn code.
func isInvalidDigits(digits []int) bool {
	const luhnMininumDigits = 2

	return len(digits) < luhnMininumDigits
}

// Filters all non-digit characters from a string and converts the digits to
// their integer value. Returns the result as a slice of integers.
func onlyDigits(id string) []int {
	digits := []int{}

	for _, character := range id {
		if unicode.IsDigit(character) {
			digits = append(digits, int(character-'0'))
		}
	}

	return digits
}

// Calculates a sum of digits used to check if a Luhn code is valid or not.
// Performs the number doubling used in Luhn every 2 digits backwards
// (starting from the second digit).
func luhnSum(digits []int) int {
	sum := 0
	digitsCount := len(digits)
	reversedPairIndexRem := digitsCount % 2

	for index := digitsCount - 1; index >= 0; index-- {
		if index%2 == reversedPairIndexRem {
			sum += doubleDigit(digits[index])
		} else {
			sum += digits[index]
		}
	}

	return sum
}

// Checks if a given "row sum" is valid. It is considered valid if it
// is a multiple of 10.
func isValidLuhnSum(sum int) bool {
	return sum%10 == 0
}

// Pattern used to check whether or not a string contains invalid characters
// for a Luhn code.
var invalidRunePattern = regexp.MustCompile(`[^\s\d]`)

// Digit doubling operation used to calculate the Luhn sum. If double the
// digit is greater than 9, subtract 9 from the value. Otherwise, return
// only double the given digit.
func doubleDigit(digit int) int {
	doubled := digit * 2

	if doubled > 9 {
		doubled = doubled - 9
	}

	return doubled
}

File: lua/luhn/luhn.lua
--- A module for validating numbers (such as credit card numbers) using the Luhn algorithm.
local luhn <const> = {}

local has_luhn_format, digits_of_luhn, luhn_sum, luhn_double

--- Validates a number string according to the Luhn algorithm.
--- The function checks if the string contains only digits and spaces, has at least two digits,
--- and if the Luhn sum is a multiple of 10.
--- @param str string The number string to validate.
--- @return boolean -- `true` if the number is valid, `false` otherwise.
function luhn.valid(str)
  if not has_luhn_format(str) then
    return false
  end

  local digits <const> = digits_of_luhn(str)

  if #digits <= 1 then
    return false
  end

  return luhn_sum(digits) % 10 == 0
end

--- Checks if a string contains only digits and spaces.
--- @param str string The string to check.
--- @return boolean -- `true` if the string has a valid format, `false` otherwise.
function has_luhn_format(str)
  return str:match("[^%d%s]") == nil
end

--- Extracts all digit characters from a string and converts them into a table of numbers.
--- @param str string The string to process.
--- @return integer[] -- A table containing the digits as numbers.
function digits_of_luhn(str)
  local digs = {}

  for digit in str:gmatch("%d") do
    table.insert(digs, tonumber(digit))
  end

  return digs
end

--- Calculates the Luhn sum of a table of digits.
--- This sum is calculated by iterating through the digits from right to left, doubling
--- every second digit, and then adding all the digits together.
--- @param digits integer[] The table of digits.
--- @return integer -- The calculated Luhn sum.
function luhn_sum(digits)
  local digits_count <const> = #digits
  local sum = 0

  for i = 1, digits_count, 1 do
    local index <const> = digits_count - i + 1
    local digit <const> = digits[index]
    if i % 2 == 0 then
      sum = sum + luhn_double(digit)
    else
      sum = sum + digit
    end
  end

  return sum
end

--- Doubles a digit as part of the Luhn algorithm.
--- If the doubled value is greater than 9, it is replaced by the sum of its digits
--- (e.g., 14 becomes 1 + 4 = 5).
--- This is equivalent to subtracting 9.
--- @param digit integer The digit to double.
--- @return integer -- The doubled digit, adjusted if greater than 9.
function luhn_double(digit)
  local result <const> = 2 * digit

  if result > 9 then
    return result - 9
  end
  return result
end

return luhn
